### PR TITLE
Customize web3mail constructor parameters

### PR DESCRIPTION
**Pull request by luisccastillo**
Modifies the web3mail constructor to take in parameters regarding the environment:
- IPFs node
- IPFS gateway
- whitelist contract address
- email sender dapp address
- workerpool address (done here since secrets are tied to the workerpol env)  
- data protector subgraph URL

the sendEmail method now allows additional parameters 
- dataset max price : set the desired max price to pay for the dataset
- workerpool max price : set the desired max price to pay for the worker
- app max price : set the desired max price to pay for the app
- label: min 3 chars max 10, used to identify the email campaign via iexec_args

### CODE REVIEW

File: src/web3mail/IExecWeb3mail.ts [hunk 1]
```diff
@@ -9,41 +9,75 @@ import {
...
 export class IExecWeb3mail {
-  fetchMyContacts: (args?: FetchContactsParams) => Promise<Contact[]>;
-  sendEmail: (args: SendEmailParams) => Promise<SendEmailResponse>;
+  private iexec: IExec;
+  private ipfsNode: string;
+  private ipfsGateway: string;
+  private dataProtectorSubgraph: string;
+  private dappAddressOrEns: AddressOrENS;
+  private dappWhitelistAddress: AddressOrENS;
+  private workerpoolAddressOrEns: AddressOrENS;
```
Comment (PierreJeanjacquot) [inline src/web3mail/IExecWeb3mail.ts:32]: `workerpoolAddressOrEns` should not be at the constructor level, only for `sendEmail` method

File: src/web3mail/types.ts [hunk 2]
```diff
@@ -22,6 +27,14 @@ export type SendEmailParams = {
   protectedData: Address;
   contentType?: string;
   senderName?: string;
+  label?: string;
+  workerpoolAddressOrEns?: AddressOrENS;
+  dappAddressOrENS?: AddressOrENS;
+  ipfsNode?: string;
+  ipfsGateway?: string;
```
Comment (PierreJeanjacquot) [inline src/web3mail/types.ts:34]: these 3 (dappAddressOrENS, ipfsNode, ipfsGateway) should not be part of the `SendEmailParams`, they are injected by the `IExecWeb3mail` instance when invoking the `sendEmail` method

you can split into specific types that will not be exposed externally (see IExecConsumer type)

```ts
export type DappAddressConsumer = {
    dappAddressOrENS: AddressOrENS;
};

export type IpfsNodeConfigConsumer = {
     ipfsNode: string;
};

export type IpfsGatewayConfigConsumer = {
    ipfsGateway: string;
};
```

File: src/web3mail/types.ts [hunk 3]
```diff
@@ -33,7 +46,20 @@ export type FetchContactsParams = {
...
+   * If not provided, the default dataProtector contract address will be used.
+   * @default{@link WEB3_MAIL_DAPP_ADDRESS}
+   */
+  dappAddressOrEns?: AddressOrENS;
+  /**
+   * The Ethereum contract address or ENS (Ethereum Name Service) for dataProtector smart contract.
+   * If not provided, the default dataProtector contract address will be used.
+   * @default{@link WHITELIST_SMART_CONTRACT_ADDRESS}
+   */
+  dappWhitelistAddress?: AddressOrENS;
```
Comment (PierreJeanjacquot) [inline src/web3mail/types.ts:60]: these (dappAddressOrENS, dappWhitelistAddress) should not be part of the `FetchContactsParams`, they are injected by the `IExecWeb3mail` instance when invoking the `fetchMyContacts` method.

you can split into specific types that will not be exposed externally (see IExecConsumer type)

File: src/web3mail/types.ts [hunk 4]
```diff
@@ -52,3 +78,55 @@ export type GraphQLResponse = {
...
+   * @default{@link WEB3_MAIL_DAPP_ADDRESS}
+   */
+  dappAddressOrEns?: AddressOrENS;
+
+  /**
+   * The workerpool used to execute the mail sender dapp
+   * If not provided, the default workerpool will be used.
+   * @default{@link WORKERPOOL_ADDRESS}
+   */
+  workerpoolAddressOrEns?: AddressOrENS;
```
Comment (PierreJeanjacquot) [inline src/web3mail/types.ts:98]: remove from the constructor part

```suggestion
```

**Review by PierreJeanjacquot — CHANGES_REQUESTED** [review]

**Review by PierreJeanjacquot — APPROVED** [review]
LGTM

**Comment by cedric25** [comment]
Waiting for: https://github.com/iExecBlockchainComputing/web3mail-sdk/pull/102

**Pull request merged**